### PR TITLE
[graph_trainer] Preserve precompile input ordering

### PR DESCRIPTION
**Pull request by bobrenjc93**
## Summary

Graph Trainer precompile flattens `extra_kwargs` into positional graph inputs. The full trace-time `TreeSpec` cannot be serialized because a FlexAttention `BlockMask` context contains the unpicklable `mask_mod` closure. Without serialization-safe binding metadata, dictionary insertion order accidentally became semantic: precompile built `attention_masks, positions`, while runtime built `positions, attention_masks`, swapping the tensors during replay.

- bind runtime dictionary values by their trace-time keys rather than insertion order
- reject stale artifacts that lack enough metadata to bind inputs safely
- align the precompile construction order with runtime as an additional safeguard
- cover reordered nested dictionaries and BlockMask artifact pickling with regression tests

## Design

`minimal_fx_tracer` builds the full pytree spec only while tracing, where it is needed to reconstruct the original `(*args, **kwargs)` call. It remains local to the trace and is not stored.

`TracedResult` and `PrecompiledFxTraceArtifact` instead share a single picklable `user_inputs_spec` for runtime binding. This spec preserves standard tuple/list/dict structure and dictionary keys while treating custom pytree nodes such as `BlockMask` as leaves, excluding their unpicklable context. Both live and precompiled replay use this same spec with `TreeSpec.flatten_up_to`.

## Behavioral contract

Keyword argument insertion order is not part of the call contract. If runtime inputs contain the same keys with compatible structure, `run_traced` accepts them in any order and reorders their flattened leaves to the trace-time key order. This matches normal Python keyword binding: values bind by name rather than by the order in which a dictionary happened to be constructed.

The runner fails only when it cannot bind safely, such as missing or extra keys, incompatible standard-container structure, or a stale precompiled artifact with no saved binding schema. In particular, merely changing `positions, attention_masks` to `attention_masks, positions` does not fail and does not swap the tensors.

## Numerics

The failing precompiled path crashes before producing a loss on main, so there is no before/after loss pair for this reproducer. This change does not alter model computation; it binds the same keyed runtime tensors to the placeholders captured for those keys.

## Test plan

- `pytest -q torchtitan/experiments/graph_trainer/tests/test_precompile.py -x` (23 passed, 1 skipped)
- `pytest -q torchtitan/experiments/graph_trainer/tests/test_trace_module.py -x` (23 passed, 34 skipped, 7 subtests passed)
- `pre-commit run --all-files`

### CODE REVIEW

File: torchtitan/experiments/graph_trainer/make_fx_tracer.py [hunk 1]
```diff
@@ -285,7 +285,11 @@ class TracedResult:
         example_inputs: Trace-time fake flat inputs used by downstream graph passes.
         num_flat_inputs: Number of flat graph inputs before subclass unwrapping.
         input_subclass_layouts: Subclass unwrap/rewrap metadata for inputs.
-        user_inputs_spec: Trace-time pytree spec for ``(args, kwargs)``.
+        user_inputs_spec: Trace-time pytree spec for ``(args, kwargs)``. This is
+            unavailable for loaded precompile artifacts whose custom pytree
+            contexts are not picklable.
+        user_inputs_order_spec: Picklable trace-time ordering for standard
+            input containers. Custom pytree nodes are treated as leaves.
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/make_fx_tracer.py]: what's the difference between these 2 fields? spec feels like an superset of order_spec. 
Can we just use user_inputs_spec?

Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/make_fx_tracer.py]: Having these 2 fields side-by-side is clear signal that this is a workaround. I think the root fix is to make user_inputs_spec carry what precompile needs, and serialize that.

Comment (bobrenjc93) [reply torchtitan/experiments/graph_trainer/make_fx_tracer.py]: The full spec now stays local to tracing, where we use it to reconstruct the original call. TracedResult and the artifact only store the serialization-safe user_inputs_spec, which treats custom nodes like BlockMask as leaves while preserving container structure and dictionary keys.
